pkgs/lib/fluentd.nix

```
@@ -0,0 +1,133 @@
+{ pkgs, cwPkgs, ... }:
+
+with builtins;
+
+rec {
+
+  enableFluentdForService = service: service ? fluentd && service.fluentd ? source && service.fluentd.source ? type;
+
+  captureServiceStdout = service:
+    service ? fluentd && service.fluentd ? source && captureSourceStdout service.fluentd.source;
+
+  captureSourceStdout = source:
+    source ? type && source.type == "stdout";
+
+  attrsToFluentd = set:
+    concatStringsSep "\n"
+      (pkgs.lib.mapAttrsToList (name: value:
+        let
+          v = if isInt value then toString value
+              else if isBool value then (if value == true then "true" else "false")
+              else if isString value then value
+              else if isAttrs value then attrsToFluentd value
+              else if isList value then map attrsToFluentd value
+              else abort "attrsToFluentd: value not supported";
+          n = if name == "type" then "@type" else name;
+          subSection = n: v: "  <${n}>\n${v}\n  </${n}>";
+        in
+          # support for fluentd 1.x
+          if isAttrs value then
+            subSection n v
+          else if isList value then
+            concatStringsSep "\n" (map (subSection n) v)
+          else
+            "  ${n} ${v}"
+      ) set);
+
+  sanitizeFluentdSource = name: source:
+    if captureSourceStdout source then
+      source // {
+        type = "named_pipe";
+        path = "/tmp/${name}";
+        # format is required
+        format = if source ? format then source.format else "none";
```

Not sure if it is required but you could throw an exception (`builtins.throw`)

---

it's just for setting a default when format is not defined, I don't want to break here